derivation/shared/src/main/scala/io/circe/derivation/Configuration.scala

```diff
@@ -0,0 +1,96 @@
+package io.circe.derivation
+
+/**
+  * Configuration allowing customisation of the JSON produced when encoding, or
+  * expected when decoding. Can be passed to the [[JsonCodec]] annotation to
+  * allow customisation of derivation.
+  *
+  * The configuration also determines if either *both* encoder and decoder are
+  * derived - or if only one of them will be.
+  */
+sealed trait Configuration[A <: Configuration[A]] {
```

Comment:
Perhaps a simpler encoding (to avoid f-bounded polymorphism) would be along these lines:
```
sealed trait GenerationStrategy
case object Both extends GenerationStrategy
// ...

case class Configuration private(
   gen: GenerationStrategy,
   transformMemberNames: String => String) {

def withSnakeCaseMemberNames // ....
}

object Configuration {

  val default = Configuration(Both, identity)
//...
}
```